perf: avoid extra copies when receiving int4 and int2 in PGStream

vishalvrv9 opened the pull request:
### All Submissions:

* [x] Have you followed the guidelines in our [Contributing](https://github.com/pgjdbc/pgjdbc/blob/master/CONTRIBUTING.md) document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes to Existing Features:

This PR is for #3250 
I have moved receiveInteger2 and receiveInteger4 to VisibleBufferedInputStream. 
Within PGStream, I have also removed dependency on int4buf and int2buf by directly writing value to OutputStream inside sendInteger2 and sendInteger4
* [x] Have you successfully run tests with your changes locally?


Based on reviews, can further add tests and documentation for the new methods

vlsi commented:
Please add benchmarks

bokken commented on pgjdbc/src/main/java/org/postgresql/core/PGStream.java:
```diff
@@ -355,11 +344,10 @@ public void sendChar(int val) throws IOException {
-    int4Buf[1] = (byte) (val >>> 16);
-    int4Buf[2] = (byte) (val >>> 8);
-    int4Buf[3] = (byte) (val);
-    pgOutput.write(int4Buf);
```
Is pgOutput always buffered internally?

vishalvrv9 replied:
based on my understanding, its always buffered except when its a char (single byte). Good observation as this can increase IO compromising performance? Is that what you were getting to? 

Can create a local buffer within this scope and write once. Not sure if this is what you were targeting?

vishalvrv9 commented:
@vlsi 

I have added benchmarks. After running the benchmarks on my machine, I'm getting consistent results as below. 

I'm pasting results of ```before change``` and ```after change``` for comparison

```
Before the PGStream refactoring

Benchmark                                      (byteArrSize)  Mode  Cnt         Score        Error  Units
StreamReceiveInteger.receiveInteger2Benchmark           4096  avgt   25     40654.447 ±    985.740  ns/op
StreamReceiveInteger.receiveInteger2Benchmark           8192  avgt   25     79654.366 ±   1004.068  ns/op
StreamReceiveInteger.receiveInteger2Benchmark          65536  avgt   25    642311.430 ±   6683.201  ns/op
StreamReceiveInteger.receiveInteger2Benchmark         131072  avgt   25   1276954.173 ±  18509.234  ns/op
StreamReceiveInteger.receiveInteger2Benchmark        1048576  avgt   25  10064545.268 ±  60657.317  ns/op

StreamReceiveInteger.receiveInteger4Benchmark           4096  avgt   25     21010.785 ±    157.727  ns/op
StreamReceiveInteger.receiveInteger4Benchmark           8192  avgt   25     40477.736 ±    926.200  ns/op
StreamReceiveInteger.receiveInteger4Benchmark          65536  avgt   25    316628.093 ±   3137.282  ns/op
StreamReceiveInteger.receiveInteger4Benchmark         131072  avgt   25    642833.953 ±   5306.471  ns/op
StreamReceiveInteger.receiveInteger4Benchmark        1048576  avgt   25   5259341.809 ± 123359.057  ns/op

After the refactoring


Benchmark                                      (byteArrSize)  Mode  Cnt        Score        Error  Units
StreamReceiveInteger.receiveInteger2Benchmark           4096  avgt   25    30660.816 ±    395.408  ns/op
StreamReceiveInteger.receiveInteger2Benchmark           8192  avgt   25    61470.316 ±   2574.740  ns/op
StreamReceiveInteger.receiveInteger2Benchmark          65536  avgt   25   482912.610 ±   4937.530  ns/op
StreamReceiveInteger.receiveInteger2Benchmark         131072  avgt   25   963792.516 ±   4081.982  ns/op
StreamReceiveInteger.receiveInteger2Benchmark        1048576  avgt   25  7761601.288 ± 147700.245  ns/op

StreamReceiveInteger.receiveInteger4Benchmark           4096  avgt   25    15744.789 ±     81.255  ns/op
StreamReceiveInteger.receiveInteger4Benchmark           8192  avgt   25    31209.710 ±    168.836  ns/op
StreamReceiveInteger.receiveInteger4Benchmark          65536  avgt   25   249877.018 ±   3850.310  ns/op
StreamReceiveInteger.receiveInteger4Benchmark         131072  avgt   25   501395.618 ±   5036.232  ns/op
StreamReceiveInteger.receiveInteger4Benchmark        1048576  avgt   25  4038478.701 ±  42636.446  ns/op
```
@bokken 
I have also changed the PGStream to buffer values before writing to OutputStream within sendInteger2() and sendInteger4()

vishalvrv9 commented:
@vlsi

I have added benchmarks. After running the benchmarks on my machine, I'm getting consistent results as below.

I'm pasting results of before change and after change for comparison

```
Before the PGStream refactoring

Benchmark                                      (byteArrSize)  Mode  Cnt         Score        Error  Units
StreamReceiveInteger.receiveInteger2Benchmark           4096  avgt   25     40654.447 ±    985.740  ns/op
StreamReceiveInteger.receiveInteger2Benchmark           8192  avgt   25     79654.366 ±   1004.068  ns/op
StreamReceiveInteger.receiveInteger2Benchmark          65536  avgt   25    642311.430 ±   6683.201  ns/op
StreamReceiveInteger.receiveInteger2Benchmark         131072  avgt   25   1276954.173 ±  18509.234  ns/op
StreamReceiveInteger.receiveInteger2Benchmark        1048576  avgt   25  10064545.268 ±  60657.317  ns/op

StreamReceiveInteger.receiveInteger4Benchmark           4096  avgt   25     21010.785 ±    157.727  ns/op
StreamReceiveInteger.receiveInteger4Benchmark           8192  avgt   25     40477.736 ±    926.200  ns/op
StreamReceiveInteger.receiveInteger4Benchmark          65536  avgt   25    316628.093 ±   3137.282  ns/op
StreamReceiveInteger.receiveInteger4Benchmark         131072  avgt   25    642833.953 ±   5306.471  ns/op
StreamReceiveInteger.receiveInteger4Benchmark        1048576  avgt   25   5259341.809 ± 123359.057  ns/op

After the refactoring


Benchmark                                      (byteArrSize)  Mode  Cnt        Score        Error  Units
StreamReceiveInteger.receiveInteger2Benchmark           4096  avgt   25    30660.816 ±    395.408  ns/op
StreamReceiveInteger.receiveInteger2Benchmark           8192  avgt   25    61470.316 ±   2574.740  ns/op
StreamReceiveInteger.receiveInteger2Benchmark          65536  avgt   25   482912.610 ±   4937.530  ns/op
StreamReceiveInteger.receiveInteger2Benchmark         131072  avgt   25   963792.516 ±   4081.982  ns/op
StreamReceiveInteger.receiveInteger2Benchmark        1048576  avgt   25  7761601.288 ± 147700.245  ns/op

StreamReceiveInteger.receiveInteger4Benchmark           4096  avgt   25    15744.789 ±     81.255  ns/op
StreamReceiveInteger.receiveInteger4Benchmark           8192  avgt   25    31209.710 ±    168.836  ns/op
StreamReceiveInteger.receiveInteger4Benchmark          65536  avgt   25   249877.018 ±   3850.310  ns/op
StreamReceiveInteger.receiveInteger4Benchmark         131072  avgt   25   501395.618 ±   5036.232  ns/op
StreamReceiveInteger.receiveInteger4Benchmark        1048576  avgt   25  4038478.701 ±  42636.446  ns/op
```

@bokken
I have also changed the PGStream to buffer values before writing to OutputStream within sendInteger2() and sendInteger4()

vlsi commented on pgjdbc/src/main/java/org/postgresql/core/PGStream.java:
```diff
@@ -372,9 +362,10 @@ public void sendInteger2(int val) throws IOException {
-    int2Buf[0] = (byte) (val >>> 8);
-    int2Buf[1] = (byte) val;
-    pgOutput.write(int2Buf);
+    byte[] buffer = new byte[2];
+    buffer[0] = (byte) (val >> 8);
+    buffer[1] = (byte) val;
+    pgOutput.write(buffer);
```
Why making this change?

vishalvrv9 replied:
to buffer bytes before writing as pgOutput is always buffered inside PGStream unless its a character. (that's how it is written everytime within PGStream

Effectively reduces the number of IO operations. Not sure if it helps, because bufferSize of pgOutput is always big.

vishalvrv9 commented on benchmarks/src/jmh/java/org/postgresql/benchmark/stream/StreamReceiveInteger.java:
```diff
@@ -0,0 +1,129 @@
+
+    QueryExecutor queryExecutor = ((PgConnection) connection).getQueryExecutor();
+
+    Field pgStreamField = queryExecutor.getClass().getSuperclass().getDeclaredField("pgStream");
```
QueryExecutorImpl was one of the classes I found using receiveInteger4() and receiveInteger2() considerably. But couldn't figure out a way to access the pgStream within QueryExecutor.

But Found something similar being done in ConnectionTest

https://github.com/pgjdbc/pgjdbc/blob/a1bd1ddfa5957b6545f5f9e84745234a434f7163/pgjdbc/src/test/java/org/postgresql/test/jdbc2/ConnectionTest.java#L533

Is there a better way to achieve this?

vlsi replied:
Have you considered `new VisibleBufferedInputStream`?

vishalvrv9 replied:
I tried something like below

```
  @Benchmark
  public void receiveInteger2Benchmark(Blackhole bh) throws Exception {
    var bufferedInputStream = new VisibleBufferedInputStream(new ByteArrayInputStream(integer2Data), 8192);

    while (bufferedInputStream.available() >= 2) {
      int value = bufferedInputStream.receiveInteger2();
      bh.consume(value);
    }
  }
```

where integer2Data = new byte[1024 * 1024]; //1 MB

But I wasn't sure how to compare this with previous implementation. As receiveInteger was being called from within PGStream.

vishalvrv9 replied:
I still have the above benchmark classes performing this benchmark (unstaged). Should I include them for review?

vlsi replied:
I do not see what is the value of reflection-based benchmarks, however, it is clear they make maintenance harder.

vishalvrv9 replied:
I've changed the benchmark to avoid reflection and directly benchmark ```VisibleBufferedInputStream()```

vishalvrv9 commented:
@vlsi 

I have refactored the benchmarks to as to avoid reflection and directly use ```VisibleBufferedInputStream```

The results are below:

```
Benchmark                                      (byteArrSize)  Mode  Cnt        Score       Error  Units
StreamReceiveInteger.receiveInteger2Benchmark           4096  avgt   25     4902.307 ±    64.914  ns/op
StreamReceiveInteger.receiveInteger2Benchmark           8192  avgt   25     9706.844 ±    24.202  ns/op
StreamReceiveInteger.receiveInteger2Benchmark          65536  avgt   25    79798.808 ±  1115.382  ns/op
StreamReceiveInteger.receiveInteger2Benchmark         131072  avgt   25   162177.605 ±  4183.496  ns/op
StreamReceiveInteger.receiveInteger2Benchmark        1048576  avgt   25  1334331.384 ± 88054.323  ns/op

StreamReceiveInteger.receiveInteger4Benchmark           4096  avgt   25     2940.606 ±    23.006  ns/op
StreamReceiveInteger.receiveInteger4Benchmark           8192  avgt   25     5951.321 ±   111.409  ns/op
StreamReceiveInteger.receiveInteger4Benchmark          65536  avgt   25    47203.798 ±   579.416  ns/op
StreamReceiveInteger.receiveInteger4Benchmark         131072  avgt   25    95262.525 ±  1138.591  ns/op
StreamReceiveInteger.receiveInteger4Benchmark        1048576  avgt   25   774245.750 ± 14734.046  ns/op
```

vlsi commented on benchmarks/src/jmh/java/org/postgresql/benchmark/stream/StreamReceiveInteger.java:
```diff
@@ -0,0 +1,92 @@
+
+  @Benchmark
+  public void receiveInteger4Benchmark(Blackhole bh) throws Exception {
+    bufferedInputStream = new VisibleBufferedInputStream(new ByteArrayInputStream(integer4Data), byteArrSize);
```
Currently we use a fixed buffer of 8192 rather than buffering all the contents always

```suggestion
    bufferedInputStream = new VisibleBufferedInputStream(new ByteArrayInputStream(integer4Data), 8192);
```

vishalvrv9 replied:
okay. The benchmark uses 8192 as the fixed buffersize now

vlsi commented on benchmarks/src/jmh/java/org/postgresql/benchmark/stream/StreamReceiveInteger.java:
```diff
@@ -0,0 +1,92 @@
+
+  @Benchmark
+  public void receiveInteger2Benchmark(Blackhole bh) throws Exception {
+    bufferedInputStream = new VisibleBufferedInputStream(new ByteArrayInputStream(integer2Data), byteArrSize);
```
```suggestion
    bufferedInputStream = new VisibleBufferedInputStream(new ByteArrayInputStream(integer2Data), 8192);
```

vlsi commented on benchmarks/src/jmh/java/org/postgresql/benchmark/stream/StreamReceiveInteger.java:
```diff
@@ -0,0 +1,92 @@
+  private VisibleBufferedInputStream bufferedInputStream;
+  @Param({"4096", "8192", "65536", "131072", "1048576"})
+  private int byteArrSize;
+  private byte[] integer2Data;
+  private byte[] integer4Data;
```
What is the purpose of having separate arrays?

vishalvrv9 replied:
No purpose. I think the idea was to separate byteArrays for the separate methods. But in retrospect a common array should be better.  Have resolved this to use a single array

vlsi commented on benchmarks/src/jmh/java/org/postgresql/benchmark/stream/StreamReceiveInteger.java:
```diff
@@ -0,0 +1,92 @@
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+public class StreamReceiveInteger {
+
+  private VisibleBufferedInputStream bufferedInputStream;
```
Please convert it to a local variable

vishalvrv9 commented:
Hey @vlsi 

Have implemented the proposed review changes. 

The benchmark results look as below:

```
Benchmark                                      Mode  Cnt      Score     Error  Units
StreamReceiveInteger.receiveInteger2Benchmark  avgt   25  12248.481 ± 315.238  ns/op
StreamReceiveInteger.receiveInteger4Benchmark  avgt   25   8068.027 ±  84.654  ns/op
```

vishalvrv9 commented:
Hi @vlsi 

I've accepted the review changes suggested by you and resolved them. Is there anything I can add to this or am I missing something?

vlsi commented on pgjdbc/src/main/java/org/postgresql/core/VisibleBufferedInputStream.java:
```diff
@@ -350,4 +350,18 @@ public void setTimeoutRequested(boolean timeoutRequested) {
     return wrapped;
   }
+
+  public int receiveInteger2() throws IOException {
+    if (ensureBytes(2)) {
+      return (buffer[index++] & 0xFF) << 8 | buffer[index++] & 0xFF;
+    }
+    throw new EOFException("End of stream reached while trying to read integer2");
+  }
+
+  public int receiveInteger4() throws IOException {
+    if (ensureBytes(4)) {
+      return (buffer[index++] & 0xFF) << 24 | (buffer[index++] & 0xFF) << 16 | (buffer[index++] & 0xFF) << 8 | buffer[index++] & 0xFF;
+    }
+    throw new EOFException("End of stream reached while trying to read integer4");
+  }
```
Would you please check out https://google.github.io/styleguide/javaguide.html#s3.4.2-ordering-class-contents ?

> new methods are not just habitually added to the end of the class

Have you considered adding the methods closer to the existing `read()` ones?

Then the naming could be like `readInt2()`, `readInt4()` (see `PgBufferedOutputStream#writeInt4`)

vlsi commented on benchmarks/src/jmh/java/org/postgresql/benchmark/stream/StreamReceiveInteger.java:
```diff
@@ -0,0 +1,82 @@
+
+  @Benchmark
+  public void receiveInteger2Benchmark(Blackhole bh) throws Exception {
+    VisibleBufferedInputStream bufferedInputStream = new VisibleBufferedInputStream(new ByteArrayInputStream(integerData),
+        integerData.length);
```
This looks like a setup to me. Have you considered moving it to `@Setup` method?

vlsi replied:
The suggestion of moving stream initialization to the setup was invalid as we can't reuse the stream, and we really have to recreate it otherwise it gets exhausted and we can no longer read from it.

vishalvrv9 commented:
Have accomodated the changes. The benchmark results are below:
```
Benchmark                                      Mode  Cnt  Score   Error  Units
StreamReceiveInteger.receiveInteger2Benchmark  avgt   25  4.829 ± 0.023  ns/op
StreamReceiveInteger.receiveInteger4Benchmark  avgt   25  4.846 ± 0.078  ns/op
```

vlsi commented on pgjdbc/src/main/java/org/postgresql/core/VisibleBufferedInputStream.java:
```diff
@@ -350,4 +364,5 @@ public void setTimeoutRequested(boolean timeoutRequested) {
   public InputStream getWrapped() {
     return wrapped;
   }
+
```
```suggestion
```

vlsi commented on benchmarks/src/jmh/java/org/postgresql/benchmark/stream/StreamReceiveInteger.java:
```diff
@@ -0,0 +1,82 @@
+
+  @Benchmark
+  public void receiveInteger2Benchmark(Blackhole bh) throws Exception {
+
```
```suggestion
```

vlsi commented on benchmarks/src/jmh/java/org/postgresql/benchmark/stream/StreamReceiveInteger.java:
```diff
@@ -0,0 +1,82 @@
+
+  @Benchmark
+  public void receiveInteger4Benchmark(Blackhole bh) throws Exception {
+
```
```suggestion
```

vlsi reviewed (approved):
Looks great to me except a couple of superfluous lines. Would you please cleanup them?
